File: backend/app/core/validators.py
"""Validadores brasileiros (CPF, CNPJ, CNS, CEP).

Todos aceitam com ou sem máscara e retornam o valor normalizado (só dígitos)
ou lançam ValueError.
"""

from __future__ import annotations

import re

_DIGITS = re.compile(r"\D")


def _only_digits(v: str) -> str:
    return _DIGITS.sub("", v)


# ─── CPF ──────────────────────────────────────────────────────────────────────


def validate_cpf(value: str) -> str:
    cpf = _only_digits(value)
    if len(cpf) != 11 or cpf == cpf[0] * 11:
        raise ValueError("CPF inválido.")

    def _dv(nums: str, weights: range) -> int:
        s = sum(int(d) * w for d, w in zip(nums, weights, strict=True))
        r = (s * 10) % 11
        return r if r < 10 else 0

    dv1 = _dv(cpf[:9], range(10, 1, -1))
    dv2 = _dv(cpf[:10], range(11, 1, -1))
    if int(cpf[9]) != dv1 or int(cpf[10]) != dv2:
        raise ValueError("CPF inválido.")
    return cpf


# ─── CNPJ ─────────────────────────────────────────────────────────────────────


def validate_cnpj(value: str) -> str:
    cnpj = _only_digits(value)
    if len(cnpj) != 14 or cnpj == cnpj[0] * 14:
        raise ValueError("CNPJ inválido.")

    def _dv(nums: str, weights: list[int]) -> int:
        s = sum(int(d) * w for d, w in zip(nums, weights, strict=True))
        r = s % 11
        return 0 if r < 2 else 11 - r

    dv1 = _dv(cnpj[:12], [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2])
    dv2 = _dv(cnpj[:13], [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2])
    if int(cnpj[12]) != dv1 or int(cnpj[13]) != dv2:
        raise ValueError("CNPJ inválido.")
    return cnpj


# ─── CNS (Cartão Nacional de Saúde) ───────────────────────────────────────────


def validate_cns(value: str) -> str:
    cns = _only_digits(value)
    if len(cns) != 15:
        raise ValueError("CNS inválido.")

    if cns[0] in {"1", "2"}:
        pis = cns[:11]
        soma = sum(int(pis[i]) * (15 - i) for i in range(11))
        dv = soma % 11
        if dv == 10:
            soma += 2
            resto = soma % 11
            dv = 11 - resto
            resultado = pis + "001" + str(dv)
        else:
            dv = 11 - dv
            resultado = pis + "000" + str(dv) if dv < 11 else pis + "000" + "0"
        if resultado != cns:
            raise ValueError("CNS inválido.")
    elif cns[0] in {"7", "8", "9"}:
        soma = sum(int(cns[i]) * (15 - i) for i in range(15))
        if soma % 11 != 0:
            raise ValueError("CNS inválido.")
    else:
        raise ValueError("CNS inválido.")
    return cns


# ─── CEP ──────────────────────────────────────────────────────────────────────


def validate_cep(value: str) -> str:
    cep = _only_digits(value)
    if len(cep) != 8:
        raise ValueError("CEP inválido.")
    return cep

File: frontend/src/lib/validators.ts
// Validadores síncronos para uso em real-time. Devolvem null se ok,
// ou uma mensagem de erro pronta para exibir.

const onlyDigits = (v: string) => v.replace(/\D/g, '')

// ─── CPF (algoritmo Mod11 oficial) ────────────────────────────────────────

export function validateCpf(cpf: string): string | null {
  const d = onlyDigits(cpf)
  if (!d) return null  // vazio é tratado pela obrigatoriedade
  if (d.length !== 11) return 'CPF deve ter 11 dígitos.'
  if (d === d[0].repeat(11)) return 'CPF inválido.'

  for (const i of [9, 10]) {
    let sum = 0
    for (let j = 0; j < i; j++) sum += parseInt(d[j], 10) * (i + 1 - j)
    let dig = (sum * 10) % 11
    if (dig === 10) dig = 0
    if (dig !== parseInt(d[i], 10)) return 'CPF inválido.'
  }
  return null
}

// ─── CNS (Cartão Nacional de Saúde) ───────────────────────────────────────
// Algoritmo oficial DATASUS — varia conforme primeiro dígito.

export function validateCns(cns: string): string | null {
  const d = onlyDigits(cns)
  if (!d) return null
  if (d.length !== 15) return 'CNS deve ter 15 dígitos.'

  const first = parseInt(d[0], 10)

  if (first === 1 || first === 2) {
    // CNS definitivo: PIS/PASEP base
    const pis = d.slice(0, 11)
    let sum = 0
    for (let i = 0; i < 11; i++) sum += parseInt(pis[i], 10) * (15 - i)
    const rem = sum % 11
    let dv = 11 - rem
    let resultado: string
    if (dv === 11) dv = 0
    if (dv === 10) {
      sum += 2
      const rem2 = sum % 11
      const dv2 = 11 - rem2
      resultado = `${pis}001${dv2 === 10 ? 0 : dv2}`
    } else {
      resultado = `${pis}000${dv}`
    }
    return resultado === d ? null : 'CNS inválido.'
  }

  if (first === 7 || first === 8 || first === 9) {
    // CNS provisório
    let sum = 0
    for (let i = 0; i < 15; i++) sum += parseInt(d[i], 10) * (15 - i)
    return sum % 11 === 0 ? null : 'CNS inválido.'
  }

  return 'CNS inválido (deve começar com 1, 2, 7, 8 ou 9).'
}

// ─── E-mail ───────────────────────────────────────────────────────────────

const EMAIL_RE = /^[^\s@]+@[^\s@]+\.[^\s@]+$/

export function validateEmail(email: string): string | null {
  if (!email) return null
  return EMAIL_RE.test(email) ? null : 'E-mail inválido.'
}

// ─── Data de nascimento ───────────────────────────────────────────────────

export function validateBirthDate(iso: string): string | null {
  if (!iso) return null
  const d = new Date(iso)
  if (Number.isNaN(d.getTime())) return 'Data inválida.'
  const today = new Date()
  if (d > today) return 'Data não pode ser futura.'
  const minYear = today.getFullYear() - 130
  if (d.getFullYear() < minYear) return `Data anterior a ${minYear}.`
  return null
}

// ─── Telefone (BR) ────────────────────────────────────────────────────────

export function validatePhone(phone: string): string | null {
  if (!phone) return null
  const d = onlyDigits(phone)
  if (d.length !== 10 && d.length !== 11) return 'Telefone deve ter 10 ou 11 dígitos.'
  return null
}

// ─── CEP ─────────────────────────────────────────────────────────────────

export function validateCep(cep: string): string | null {
  if (!cep) return null
  const d = onlyDigits(cep)
  if (d.length !== 8) return 'CEP deve ter 8 dígitos.'
  return null
}
